### PR TITLE
Generate ccache wrapper scripts instead of symlink

### PR DESCRIPTION
**Pull request by chenhao94**
Tested with Bazel 5.2. It is a workaround for issue
https://github.com/bazelbuild/rules_cc/issues/130

### CODE REVIEW

File: scripts/toolchain.sh [hunk 1]
```diff
@@ -14,10 +14,16 @@ pushd "$(dirname "$0")/.."
...
     rm -rf 'run/toolchain'
     mkdir -p 'run/toolchain'
-    ln -sf "$(which ccache)" "run/toolchain/$CC"
-    ln -sf "$(which ccache)" "run/toolchain/$CXX"
-    export CC="$( realpath -e "run/toolchain")/$CC"
-    export CXX="$(realpath -e "run/toolchain")/$CXX"
+    NEW_CC_DIR="$( realpath -e "run/toolchain")"
+
+    for compiler in $CC $CXX; do
+        echo """#!/bin/bash
```
Comment (xkszltl) [inline scripts/toolchain.sh:20]: 1. What's """?
2. quotes
3. cat

```
cat << EOF
#!/bin/bash
'$(which ccache)' '$(which "$compiler")' \$@
EOF
```

**Comment by xkszltl** [comment]
- What's the bug previously?
- Is it fixed in upstream?
- If it's symlink-related, can we use the same CC approach with scripts instead of symlink?
- Put some comments in place about what's working and what's not.

**Comment by chenhao94** [comment]
> * What's the bug previously?

ccache symlink to clang will be derefed by Bazel

> * Is it fixed in upstream?

No.

> * If it's symlink-related, can we use the same CC approach with scripts instead of symlink?

That is exactly what this PR does.

> * Put some comments in place about what's working and what's not.

Will do

**Comment by xkszltl** [comment]
> That is exactly what this PR does.

I mean give it CC without PATH

File: scripts/toolchain.sh [hunk 2]
```diff
@@ -14,10 +14,18 @@ pushd "$(dirname "$0")/.."
 if which ccache >/dev/null 2>&1 && ([ -d 'run' ] && [ -w 'run' ] || [ -w '.' ]); then
     rm -rf 'run/toolchain'
     mkdir -p 'run/toolchain'
-    ln -sf "$(which ccache)" "run/toolchain/$CC"
-    ln -sf "$(which ccache)" "run/toolchain/$CXX"
-    export CC="$( realpath -e "run/toolchain")/$CC"
-    export CXX="$(realpath -e "run/toolchain")/$CXX"
+    export CR_CCACHE_CC_DIR="$( realpath -e "run/toolchain")"
```
Comment (xkszltl) [inline scripts/toolchain.sh:17]: remove space, that was for alignment.

File: scripts/toolchain.sh [hunk 3]
```diff
@@ -14,10 +14,18 @@ pushd "$(dirname "$0")/.."
...
+
+    for compiler in $CC $CXX; do
+        # Bazel may dereference the ccache symlink, so we use wrapper scripts instead.
+        # See https://github.com/bazelbuild/rules_cc/issues/130
+        cat << EOF > "$CR_CCACHE_CC_DIR/$compiler"
```
Comment (xkszltl) [inline scripts/toolchain.sh:22]: You can indent EOF with underscores, and script with spaces.
```
cat <<< ____EOF | sed 's/^    //' > ...
    xxxxxxxx
____EOF
```

There's also a normal way to tell cat to ignore spaces before EOF, but I had some issue with that somewhere so rarely bother to try.

File: scripts/codebase_utils.sh [hunk 4]
```diff
@@ -13,5 +13,7 @@ export CRIS_CODE_DIRS="$(sed 's/^[[:space:]]*//' <<< "
 " | grep '[^[:space:]]')"
 
 function bazel_common_all () {
-    bazel "$@" '//...'
+    # Need to set PATH instead of CC because Bazel does not
```
Comment (xkszltl) [inline scripts/codebase_utils.sh:16]: To me there's really no need to have a line break here.

File: scripts/toolchain.sh [hunk 5]
```diff
@@ -14,10 +14,18 @@ pushd "$(dirname "$0")/.."
...
+#!/bin/bash
+$(which ccache) $(which $compiler) \$@
```
[the remark this answers is not in the document]
Comment (chenhao94) [reply scripts/toolchain.sh:24]: On which one? I guess we don't need to quote $@, right?

[the remark this answers is not in the document]
Comment (xkszltl) [reply scripts/toolchain.sh:24]: Every $ in this line, including $@

[the remark this answers is not in the document]
Comment (xkszltl) [reply scripts/toolchain.sh:24]: $compiler

**Comment by xkszltl** [comment]
> ... or define a Bazel toolchain. The latter one seems to be a preferred way.

Not sure what does toolchain looks like here, but it should still use the full path.
Problem with PATH is you're not suppose to expose this to whatever Bazel's calling.
They may not like to PATH required by PATH.
PATH is used by everything, but things like CC has a much narrower context.

File: scripts/toolchain.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (chenhao94) [reply scripts/toolchain.sh:24]: > - `$()` should be single-quoted.

I don't want $() to be in the file

[the remark this answers is not in the document]
Comment (xkszltl) [reply scripts/toolchain.sh:24]: That's why you need single-quote.
`cat <<` is equivalent to double quote, so your $() will always be expanded. But the result can have spaces, so you need to have single quote around. The single quote doesn't store the expansion of `cat <<`, it's just text.

[the remark this answers is not in the document]
Comment (xkszltl) [reply scripts/toolchain.sh:24]: You're suppose to end up with `'/usr/bin/ccache' '/usr/bin/gcc' "$@"`.

File: scripts/toolchain.sh [hunk 6]
```diff
@@ -14,10 +14,17 @@ pushd "$(dirname "$0")/.."
...
-    export CC="$( realpath -e "run/toolchain")/$CC"
-    export CXX="$(realpath -e "run/toolchain")/$CXX"
+    export CR_CCACHE_CC_DIR="$(realpath -e "run/toolchain")"
+
+    for compiler in "$CC"; do
+        # Bazel may dereference the ccache symlink, so we use wrapper scripts instead.
+        # See https://github.com/bazelbuild/rules_cc/issues/130
+        cat << ____________EOF | sed 's/^            //' > "$CR_CCACHE_CC_DIR/$(basename "$compiler")"
+            #!/bin/bash
+            '$(which ccache)' '$(which $compiler)' "\$@"
```
Comment (xkszltl) [inline scripts/toolchain.sh:24]: double quote $compiler

File: scripts/toolchain.sh [hunk 7]
```diff
@@ -14,10 +14,17 @@ pushd "$(dirname "$0")/.."
 if which ccache >/dev/null 2>&1 && ([ -d 'run' ] && [ -w 'run' ] || [ -w '.' ]); then
     rm -rf 'run/toolchain'
     mkdir -p 'run/toolchain'
-    ln -sf "$(which ccache)" "run/toolchain/$CC"
-    ln -sf "$(which ccache)" "run/toolchain/$CXX"
-    export CC="$( realpath -e "run/toolchain")/$CC"
-    export CXX="$(realpath -e "run/toolchain")/$CXX"
+    export CR_CCACHE_CC_DIR="$(realpath -e "run/toolchain")"
```
Comment (xkszltl) [inline scripts/toolchain.sh:17]: Do you really need to export this?

File: scripts/toolchain.sh [hunk 8]
```diff
@@ -14,16 +14,19 @@ pushd "$(dirname "$0")/.."
 if which ccache >/dev/null 2>&1 && ([ -d 'run' ] && [ -w 'run' ] || [ -w '.' ]); then
     rm -rf 'run/toolchain'
     mkdir -p 'run/toolchain'
-    export CR_CCACHE_CC_DIR="$(realpath -e "run/toolchain")"
+    CR_CCACHE_CC_DIR="$(realpath -e "run/toolchain")"
 
-    for compiler in "$CC"; do
+    for compiler_varname in "CC"; do
+        compiler=$(eval echo "\$$compiler_varname")
```
Comment (xkszltl) [inline scripts/toolchain.sh:20]: double quote...

**Review by xkszltl — APPROVED** [review]

**Pull request merged**